src/ArrayCache.php

```diff
@@ -47,32 +48,43 @@ public function get($key, $default = null)
             return Promise\resolve($default);
         }
 
+        if (isset($this->expires[$key]) && $this->expires[$key] < time()) {
+            unset($this->data[$key], $this->expires[$key]);
+            return Promise\resolve();
+        }
+
         // remove and append to end of array to keep track of LRU info
         $value = $this->data[$key];
         unset($this->data[$key]);
         $this->data[$key] = $value;
-
-        return Promise\resolve($value);
+        return Promise\resolve($this->data[$key]);
     }
 
-    public function set($key, $value)
+    public function set($key, $value, $ttl = null)
     {
+        $expires = null;
+
+        if (is_int($ttl)) {
+            $this->expires[$key] = time() + $ttl;
+        }
+
         // unset before setting to ensure this entry will be added to end of array
         unset($this->data[$key]);
         $this->data[$key] = $value;
 
         // ensure size limit is not exceeded or remove first entry from array
         if ($this->limit !== null && count($this->data) > $this->limit) {
             reset($this->data);
-            unset($this->data[key($this->data)]);
+            $key = key($this->data);
+            unset($this->data[$key], $this->expires[$key]);
```

Comment:
This will need to check if any items are expired first and then remove this one. Only if no expired items can be found, it should overwrite the oldest item (LRU from #26).

Right now, this may overwrite an item that an item that is not expired already.

For practical reasons it may overwrite *any* expired item. We can either loop over this array with `O(n)` or use a `SplPriorityQueue` with `O(log n)` which would make this faster for larger arrays (premature optimization?).

Reply:
Good point I'll fix that :+1: .

I'de vote for making it working now and do the `SplPriorityQueue` optimization in a follow up PR with benchmarks to see how big the difference is for different cache sizes.

Reply:
Took me a bit, but got it to work with `SplPriorityQueue` pushing soon :+1: